contracts/SpokePool.sol

```diff
@@ -1111,28 +1337,73 @@ abstract contract SpokePool is
         uint256 amountToReturn,
         uint256[] memory refundAmounts,
         uint32 leafId,
-        address l2TokenAddress,
-        address[] memory refundAddresses
-    ) internal {
-        if (refundAddresses.length != refundAmounts.length) revert InvalidMerkleLeaf();
-
-        // Send each relayer refund address the associated refundAmount for the L2 token address.
-        // Note: Even if the L2 token is not enabled on this spoke pool, we should still refund relayers.
-        uint256 length = refundAmounts.length;
-        for (uint256 i = 0; i < length; ++i) {
-            uint256 amount = refundAmounts[i];
-            if (amount > 0) IERC20Upgradeable(l2TokenAddress).safeTransfer(refundAddresses[i], amount);
+        bytes32 l2TokenAddress,
+        bytes32[] memory refundAddresses
+    ) internal returns (bool deferredRefunds) {
+        uint256 numRefunds = refundAmounts.length;
+        if (refundAddresses.length != numRefunds) revert InvalidMerkleLeaf();
+
+        if (numRefunds > 0) {
+            uint256 spokeStartBalance = IERC20Upgradeable(l2TokenAddress.toAddress()).balanceOf(address(this));
+            uint256 totalRefundedAmount = 0; // Track the total amount refunded.
+
+            // Send each relayer refund address the associated refundAmount for the L2 token address.
+            // Note: Even if the L2 token is not enabled on this spoke pool, we should still refund relayers.
+            for (uint256 i = 0; i < numRefunds; ++i) {
+                if (refundAmounts[i] > 0) {
+                    totalRefundedAmount += refundAmounts[i];
+
+                    // Only if the total refunded amount exceeds the spoke starting balance, should we revert. This
+                    // ensures that bundles are atomic, if we have sufficient balance to refund all relayers and
+                    // prevents can only re-pay some of the relayers.
+                    if (totalRefundedAmount > spokeStartBalance) revert InsufficientSpokePoolBalanceToExecuteLeaf();
+
+                    bool success = _noRevertTransfer(
```

Comment:
@pxrl how much lift do you think there will be for the executor to deal with this change where refunds can now fail?

Reply:
note that they dont fail. the behaviour is _more_ kind to the executor. in a very rare event where you set a blacklisted recipient in the transfer then the recipient gets a credit they can claim. the design here should require _no additional changes to the executor_. we should, however, listen to events that indicate if this has happened.